File: src/types/dossie.ts
/**
 * Mirror of `src-tauri/src/models/dossie.rs` (MVP 3 — Dossiê Operacional).
 * Field names stay snake_case so serde's default render matches the wire.
 *
 * Each row has a `raw_json` string preserving the verbatim mobile payload
 * for forward-compat. The structured columns are what the UI uses.
 */

import type { Import } from "./import";
import type { Occurrence } from "./occurrence";

export interface ChecklistItem {
  id: string;
  occurrence_id: string;
  import_id: string;
  original_id: string | null;
  category: string | null;
  question: string;
  required: boolean;
  /** `sim` | `nao` | `nao_se_aplica` | `nao_verificado` */
  answer: string;
  note: string | null;
  default_note: string | null;
  /** `base` | `adicionado` */
  origin: string;
  sort_order: number;
  raw_json: string;
  created_at: string;
}

export interface ChecklistSummary {
  total: number;
  answered: number;
  not_verified: number;
  not_applicable: number;
  required_total: number;
  required_pending: number;
}

export type EntityType = "vehicle" | "victim";

export interface Entity {
  id: string;
  occurrence_id: string;
  import_id: string;
  original_id: string | null;
  type: EntityType;
  identifier: string | null;
  label: string | null;
  summary: string | null;
  /** JSON-encoded array of strings. */
  photo_ids_json: string;
  raw_json: string;
  sort_order: number;
  created_at: string;
}

export interface Trace {
  id: string;
  occurrence_id: string;
  import_id: string;
  original_id: string | null;
  identifier: string | null;
  type: string | null;
  description: string | null;
  location_description: string | null;
  length: number | null;
  width: number | null;
  unit: string | null;
  direction: string | null;
  note: string | null;
  photo_ids_json: string;
  sketch_element_ids_json: string;
  raw_json: string;
  sort_order: number;
  created_at: string;
}

export interface Measurement {
  id: string;
  occurrence_id: string;
  import_id: string;
  original_id: string | null;
  label: string | null;
  point_a: string | null;
  point_b: string | null;
  value: number | null;
  unit: string | null;
  method: string | null;
  note: string | null;
  photo_ids_json: string;
  sketch_element_ids_json: string;
  raw_json: string;
  sort_order: number;
  created_at: string;
}

export interface FieldNote {
  id: string;
  occurrence_id: string;
  import_id: string;
  original_id: string | null;
  text: string | null;
  category: string | null;
  priority: string | null;
  note_created_at: string | null;
  note_updated_at: string | null;
  raw_json: string;
  sort_order: number;
  created_at: string;
}

export interface TimelineEvent {
  id: string;
  occurrence_id: string;
  import_id: string;
  original_id: string | null;
  type: string | null;
  title: string | null;
  description: string | null;
  occurred_at: string | null;
  raw_json: string;
  sort_order: number;
  created_at: string;
}

export interface OccurrenceStats {
  id: string;
  occurrence_id: string;
  import_id: string;
  duration_seconds: number | null;
  photos_count: number | null;
  victims_count: number | null;
  vehicles_count: number | null;
  traces_count: number | null;
  measurements_count: number | null;
  notes_count: number | null;
  checklist_items_count: number | null;
  answered_checklist_items_count: number | null;
  not_applicable_items_count: number | null;
  best_gps_accuracy_m: number | null;
  gps_readings_count: number | null;
  raw_json: string;
  created_at: string;
}

export interface DossieCounts {
  photos: number;
  vehicles: number;
  victims: number;
  traces: number;
  measurements: number;
  notes: number;
  timeline: number;
  checklist: ChecklistSummary;
}

export interface DossieSummary {
  occurrence: Occurrence;
  latest_import: Import | null;
  stats: OccurrenceStats | null;
  counts: DossieCounts;
}

export interface RehydrateOutcome {
  rehydrated: boolean;
  from_package_path: string | null;
  checklist_loaded: number;
  entities_loaded: number;
  traces_loaded: number;
  measurements_loaded: number;
  notes_loaded: number;
  timeline_loaded: number;
  stats_loaded: boolean;
  warnings: string[];
}

File: src-tauri/src/models/dossie.rs
//! Dossiê Operacional models (MVP 3).
//!
//! Each struct mirrors one of the structured tables introduced in
//! `migrations/005_dossie.sql`. Field names stay snake_case so serde's
//! default rendering matches the TypeScript wire format.
//!
//! Convention for "raw payload" preservation: every row stores a
//! `raw_json` string containing the original mobile JSON object verbatim.
//! The structured columns above are what the Desktop UI *uses*; the raw
//! payload is what the Desktop *never loses*.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

// ---------------------------------------------------------------------------
// checklist

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChecklistItem {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    pub import_id: Uuid,
    pub original_id: Option<String>,
    pub category: Option<String>,
    pub question: String,
    pub required: bool,
    /// `sim` | `nao` | `nao_se_aplica` | `nao_verificado`
    pub answer: String,
    pub note: Option<String>,
    pub default_note: Option<String>,
    /// `base` | `adicionado`
    pub origin: String,
    pub sort_order: i32,
    pub raw_json: String,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Default)]
pub struct ChecklistSummary {
    pub total: u32,
    pub answered: u32,
    pub not_verified: u32,
    pub not_applicable: u32,
    pub required_total: u32,
    pub required_pending: u32,
}

// ---------------------------------------------------------------------------
// entities (vehicle / victim — polymorphic)

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Entity {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    pub import_id: Uuid,
    pub original_id: Option<String>,
    /// `vehicle` | `victim`
    pub r#type: String,
    pub identifier: Option<String>,
    pub label: Option<String>,
    pub summary: Option<String>,
    /// JSON array of original photo IDs (strings).
    pub photo_ids_json: String,
    pub raw_json: String,
    pub sort_order: i32,
    pub created_at: DateTime<Utc>,
}

// ---------------------------------------------------------------------------
// traces (vestígios)

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Trace {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    pub import_id: Uuid,
    pub original_id: Option<String>,
    pub identifier: Option<String>,
    pub r#type: Option<String>,
    pub description: Option<String>,
    pub location_description: Option<String>,
    pub length: Option<f64>,
    pub width: Option<f64>,
    pub unit: Option<String>,
    pub direction: Option<String>,
    pub note: Option<String>,
    pub photo_ids_json: String,
    pub sketch_element_ids_json: String,
    pub raw_json: String,
    pub sort_order: i32,
    pub created_at: DateTime<Utc>,
}

// ---------------------------------------------------------------------------
// measurements

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Measurement {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    pub import_id: Uuid,
    pub original_id: Option<String>,
    pub label: Option<String>,
    pub point_a: Option<String>,
    pub point_b: Option<String>,
    pub value: Option<f64>,
    pub unit: Option<String>,
    pub method: Option<String>,
    pub note: Option<String>,
    pub photo_ids_json: String,
    pub sketch_element_ids_json: String,
    pub raw_json: String,
    pub sort_order: i32,
    pub created_at: DateTime<Utc>,
}

// ---------------------------------------------------------------------------
// field_notes

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FieldNote {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    pub import_id: Uuid,
    pub original_id: Option<String>,
    pub text: Option<String>,
    pub category: Option<String>,
    pub priority: Option<String>,
    pub note_created_at: Option<DateTime<Utc>>,
    pub note_updated_at: Option<DateTime<Utc>>,
    pub raw_json: String,
    pub sort_order: i32,
    pub created_at: DateTime<Utc>,
}

// ---------------------------------------------------------------------------
// timeline_events

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TimelineEvent {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    pub import_id: Uuid,
    pub original_id: Option<String>,
    pub r#type: Option<String>,
    pub title: Option<String>,
    pub description: Option<String>,
    pub occurred_at: Option<DateTime<Utc>>,
    pub raw_json: String,
    pub sort_order: i32,
    pub created_at: DateTime<Utc>,
}

// ---------------------------------------------------------------------------
// occurrence_stats

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OccurrenceStats {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    pub import_id: Uuid,
    pub duration_seconds: Option<i64>,
    pub photos_count: Option<i64>,
    pub victims_count: Option<i64>,
    pub vehicles_count: Option<i64>,
    pub traces_count: Option<i64>,
    pub measurements_count: Option<i64>,
    pub notes_count: Option<i64>,
    pub checklist_items_count: Option<i64>,
    pub answered_checklist_items_count: Option<i64>,
    pub not_applicable_items_count: Option<i64>,
    pub best_gps_accuracy_m: Option<f64>,
    pub gps_readings_count: Option<i64>,
    pub raw_json: String,
    pub created_at: DateTime<Utc>,
}

// ---------------------------------------------------------------------------
// Aggregated dossier summary returned by `get_dossie_summary`.

#[derive(Debug, Clone, Serialize)]
pub struct DossieSummary {
    pub occurrence: super::Occurrence,
    pub latest_import: Option<super::Import>,
    pub stats: Option<OccurrenceStats>,
    pub counts: DossieCounts,
}

#[derive(Debug, Clone, Serialize, Default)]
pub struct DossieCounts {
    pub photos: u32,
    pub vehicles: u32,
    pub victims: u32,
    pub traces: u32,
    pub measurements: u32,
    pub notes: u32,
    pub timeline: u32,
    pub checklist: ChecklistSummary,
}

#[derive(Debug, Clone, Serialize, Default)]
pub struct RehydrateOutcome {
    pub rehydrated: bool,
    pub from_package_path: Option<String>,
    pub checklist_loaded: u32,
    pub entities_loaded: u32,
    pub traces_loaded: u32,
    pub measurements_loaded: u32,
    pub notes_loaded: u32,
    pub timeline_loaded: u32,
    pub stats_loaded: bool,
    pub warnings: Vec<String>,
}
